### PR TITLE
[functionalization] Regenerate one output of a multi-output view without replaying the op

### PR DESCRIPTION
**Pull request by anijain2305**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.17.0) (oldest at bottom):
* __->__ #194037

Functionalization records a ViewMeta per view so it can rebuild that view after
its base is mutated. For a multi-output view the recorded replay is the whole
operation plus an index, so one output of unbind is rebuilt as
`at::unbind(base, dim)[i]`. Rebuilding all N outputs of one base allocates N*N
tensors, and under torch.compile N*N fake tensors and proxies, leaving one
redundant split node in the post-AOT graph per regeneration.

split.Tensor, split_with_sizes and unbind.int are the only multi-output views,
and each output is exactly one slice or select of the base. Codegen now emits
`forward` for those three to build that view directly, leaving a single split
node whatever N is.

The one thing replaying the operation provided that a select does not is
autograd's refusal to let you mutate an output of a multi-output view. That
restriction is keyed off CreationMeta, not off the grad_fn, so
apply_view_meta_sequence restores it directly by marking each regenerated
multi-output view MULTI_OUTPUT_NODE. Only DEFAULT is overwritten, since a view
created under no-grad or inference mode already carries a more specific reason
to reject the mutation. ATen cannot reach CreationMeta directly, because the
internal build splits aten from torch/csrc, hence the new
VariableHooksInterface method.

The cost is that a regenerated view reports SelectBackward0 where eager reports
UnbindBackward0, and the error message loses the output index. The gradient is
unchanged: select_backward scatters into the slice unbind_backward would have.
Two tests asserted on the grad_fn name and now assert the mutation error, which
is what the name stood for.

Measured against the parent, cold compile, fresh inductor cache, median of 3.
The repro mutates N split() views of one tensor and reads them back:

    def f(x):
        outs = list(x.split(1))
        for i, o in enumerate(outs):
            o.add_(i)
        return torch.cat([o.clone() for o in outs])

    N=128:  32.4s -> 14.9s
    N=275: 109.8s -> 40.1s

The post-AOT graph goes from 32 split nodes to 1 at N=16. Without the fix the
cost grows faster than N (2.15x the views, 3.4x the time); with it, 2.7x.

Test Plan:

```
python -m pytest -q -p no:randomly test/functorch/test_eager_transforms.py \
  test/test_functionalization.py test/inductor/test_auto_functionalize.py
# 875 passed, 8 skipped, 14 xfailed

python -m pytest -q -p no:randomly test/functorch/test_aotdispatch.py
# 2884 passed, 819 skipped, 55 xfailed
# (1 pre-existing failure needing LAPACK, unrelated)
```

The changed goldens are all regeneration sites, where a split or unbind plus its
dead getitems becomes a single slice or select.

This commit was authored with the assistance of an AI agent (Claude Code).

cc @ezyang @gchanan @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo

### CODE REVIEW

File: torch/csrc/functorch/init.cpp [hunk 1]
```diff
@@ -186,15 +186,16 @@ static Tensor _unwrap_functional_tensor(
...
-  // the input is up to date by committing any pending updates to the alias.
+  // Multi-output views returned to the user must replay the original operation
+  // to preserve autograd's mutation restrictions, even if an internal sync
+  // already regenerated the value using a select or slice.
   at::functionalization::impl::FunctionalizationReapplyViewsGuard guard(
       add_back_views);
-  bool any_updates = functional->apply_updates();
-  if (any_updates) {
+  functional->apply_updates();
+  if (add_back_views && functional->is_multi_output_view()) {
```
Comment (aorenste) [inline torch/csrc/functorch/init.cpp]: [blocker] My agent says:
> This branch is not gated on staleness, so it replays the whole view op every time - including when nothing was mutated, where the old `if (any_updates)` did nothing at all. Emulating it on the base build (`apply_view_meta_sequence` per output) for `torch.func.functionalize(lambda x: list(x.unbind(0)))`: at N=16 the traced graph goes from 1 `aten.unbind.int` node to 17, and eager wall time goes 0.17 ms -> 0.55 ms; at N=128, 0.83 -> 16.7 ms; at N=256, 1.58 -> 62.5 ms. The ratio doubles as N doubles - the same O(N^2) this PR removes, reintroduced at the unwrap boundary. It also hands `is_fake()` and `maybe_get_fake_mode()` (`torch/_subclasses/fake_tensor.py:286`, `:321`) a tensor-allocating side effect, since both call `_unwrap_functional_tensor`.
>
> The unconditional replay is only needed when `value_` came from a cheap `sync_`, so track that instead:
>
> ```cpp
>   functional->apply_updates();
>   const bool exact = add_back_views && functional->is_multi_output_view();
>   if (!functional->is_up_to_date() ||
>       (exact && functional->regenerated_single_output())) {
>     functional->regenerate_from_base(/*single_output_replay=*/!exact);
>   }
> ```
>
> with `regenerated_single_output_` set in `regenerate_from_base` and cleared by a non-lazy `replace_`.

File: torch/csrc/functorch/init.cpp [hunk 2]
```diff
@@ -186,15 +186,16 @@ static Tensor _unwrap_functional_tensor(
   auto functional =
       at::functionalization::impl::unsafeGetFunctionalWrapper(self);
 
-  // when regenerating the (potentially mutated) input tensors, the
-  // functionalization pass regenerates them through a series of view_copy() op
-  // calls. Functorch wants to turn those back into view ops though. Ensure that
-  // the input is up to date by committing any pending updates to the alias.
+  // Multi-output views returned to the user must replay the original operation
```
Comment (aorenste) [inline torch/csrc/functorch/init.cpp]: [cleanup] My agent says:
> This replaced the only explanation of the `FunctionalizationReapplyViewsGuard` on the next line ("...regenerates them through a series of view_copy() op calls. Functorch wants to turn those back into view ops though") and of the `apply_updates()` call, and it overstates the code: multi-output views only replay the original op when `add_back_views` is true. With `remove='mutations_and_views'` the tensor falls through to the cheap branch and its `grad_fn` becomes `SelectBackward0_copy` instead of `UnbindBackward0_copy`. That arm is safe - `*_copy` outputs are not views (`_base is None`, in-place allowed) and `select_copy` inherits `select_backward`, so gradients are identical - but nothing says so, and no test covers it.
>
> ```cpp
>   // Regeneration goes through view_copy() ops; functorch wants real views
>   // back, which is what the guard below selects. A multi-output view handed
>   // to the user must replay the original op to keep autograd's mutation
>   // restriction, even if an internal sync already rebuilt the value with a
>   // select or slice. Only under add_back_views: with view_copy the outputs
>   // are not views, so that restriction does not exist.
> ```

File: torchgen/gen_functionalization_type.py [hunk 3]
```diff
@@ -944,6 +979,29 @@ def opcall(self, is_reverse: bool, reapply_views: bool) -> str:
 
         return f"{opname}({arguments}){maybe_index}"
 
+    @property
+    def single_output_replay_body(self) -> str | None:
+        return SINGLE_OUTPUT_VIEW_REPLAY.get(str(self.f.func.name))
```
Comment (aorenste) [inline torchgen/gen_functionalization_type.py]: [cleanup] My agent says:
> Nothing checks these keys against the ops actually generated, so a fourth multi-output view op with a `view_copy` counterpart silently inherits the base `forward_single_output` and stays quadratic - and unlike a rename, it has no golden to flag it. Enumerating today's schemas gives exactly `split.Tensor`, `split_with_sizes`, `unbind.int` as the non-composite multi-output view groups, so this fires on nothing now:
>
> ```python
>     @property
>     def single_output_replay_body(self) -> str | None:
>         name = str(self.f.func.name)
>         if self.is_multi_output and name not in SINGLE_OUTPUT_VIEW_REPLAY:
>             raise AssertionError(f"multi-output view {name} has no single-output replay")
>         return SINGLE_OUTPUT_VIEW_REPLAY.get(name)
> ```

File: torch/_functorch/eager_transforms.py [hunk 4]
```diff
@@ -1586,7 +1586,10 @@ def _maybe_unwrap_functional_tensor(maybe_tensor: Any, *, reapply_views: bool) -
         # This can happen if we functionalize a fn that returns a global,
         # which was never wrapped properly.
         return maybe_tensor
-    # Sync any pending updates on the output tensor
+    # Not needed: _unwrap_functional_tensor syncs as part of deciding how to
+    # regenerate the value. It is not free either - for a multi-output view the
+    # sync's cheap replay is traced before the exact one replaces it, leaving a
+    # dead view node in the graph.
     torch._sync(maybe_tensor)
```
Comment (aorenste) [inline torch/_functorch/eager_transforms.py:1590]: [cleanup] My agent says:
> The comment argues this call is unnecessary and actively harmful, and the call is still here. The claim checks out against the new C++: `_unwrap_functional_tensor`'s new `else if (!functional->is_up_to_date())` branch is exactly `sync_()`'s own predicate, so the line can go and the comment with it, leaving:
>
> ```python
>     return _unwrap_functional_tensor(maybe_tensor, reapply_views)
> ```
>
> Keeping it costs what the comment describes - for a multi-output view the cheap `select`/`slice` is traced and then overwritten by the exact replay, leaving a dead node the base build does not emit.

**Review by aorenste — DISMISSED** [review]
(Reviewed by me, assisted by AI)

[blocker] My agent says:
> `FunctionalTensorWrapper::set__impl` (`aten/src/ATen/FunctionalTensorWrapper.cpp:257-262`) copies `value_`, `generation_`, `view_metas_` and `is_symbolic_` from `other`, but not `is_multi_output_view_`:
>
> ```cpp
>   is_symbolic_ = other->is_symbolic_;
>   is_multi_output_view_ = other->is_multi_output_view_;   // missing
> ```
>
> That omission was harmless while the flag only fed AOTAutograd's output classification. This PR makes it decide whether a user-visible tensor gets the exact replay or the cheap one, so after `a.set_(b)` with `b` an `unbind`/`split` output, `a` holds a multi-output `ViewMeta` chain while `is_multi_output_view()` reports false, and `_unwrap_functional_tensor` hands back a `select` where autograd expects an `unbind`. No repro built - `set_` under functionalization is heavily restricted - but the invariant the new guard leans on, that the flag agrees with the chain, is not maintained anywhere.

[cleanup] My agent says:
> `split.Tensor`'s short-final-chunk arm is untested. The generated body leans on `slice` clamping the end, and every split in the functionalization tests divides evenly: `test/test_functionalization.py` uses `(4,2).split(2)` and reads `split_with_sizes([2,2])` back at index 0, so the accumulation loop never runs an iteration either. One uneven case next to the new test covers the clamp and the offsets against an eager reference (verified to pass on the base build; chunks come out `(2,3) (2,3) (1,3)`):
>
> ```python
> def test_multioutput_view_regeneration_matches_eager(self, device):
>     def f(x):
>         base = x.clone()
>         outs = base.split(2)
>         base.add_(1)
>         return tuple(o.clone() for o in outs)
>
>     x = torch.arange(15.0, device=device).reshape(5, 3)
>     self.assertEqual(torch.func.functionalize(f)(x), f(x))
> ```

[cleanup] My agent says:
> The description says AOTAutograd replays the real op "through `_functionalize_apply_view_metas`", but that attribute does not exist - `hasattr(torch, "_functionalize_apply_view_metas")` is False, and the only occurrence in the tree is a stale stub in `tools/pyi/gen_pyi.py:1254`. The path that actually does it is `torch._C._functionalization.apply_view_meta_sequence`, called from `torch/_functorch/_aot_autograd/functional_utils.py:343` with the default `single_output_replay=false`. The argument holds; the name doesn't.

**Comment by anijain2305** [comment]
@claude review this PR

**Comment by anijain2305** [comment]
@claude review this PR

File: torch/_functorch/eager_transforms.py [hunk 5]
```diff
@@ -1586,8 +1586,6 @@ def _maybe_unwrap_functional_tensor(maybe_tensor: Any, *, reapply_views: bool) -
         # This can happen if we functionalize a fn that returns a global,
         # which was never wrapped properly.
         return maybe_tensor
-    # Sync any pending updates on the output tensor
-    torch._sync(maybe_tensor)
```
Comment (zou3519) [inline torch/_functorch/eager_transforms.py:1590]: this is suspicious

File: aten/src/ATen/FunctionalStorageImpl.h [hunk 6]
```diff
@@ -107,6 +107,20 @@ struct ViewMeta {
...
+  // output instead of replaying the whole operation and discarding the sibling
+  // views (e.g. one output of `unbind` is a `select` of the base). Replaying is
+  // O(number of outputs), so regenerating every view of one base is quadratic.
+  //
+  // Only valid where the result is not visible to autograd. The cheap form
+  // loses the multi-output-view bookkeeping that replaying the real op sets up:
+  // an `unbind` output regenerated as a `select` gets SelectBackward rather
+  // than UnbindBackward, and so does not inherit the restrictions autograd
+  // places on mutating a multi-output view.
+  virtual Tensor forward_single_output(const Tensor& base) {
```
Comment (zou3519) [inline aten/src/ATen/FunctionalStorageImpl.h]: so this PR generally makes sense to me but the naming is a little weird

Comment (zou3519) [reply aten/src/ATen/FunctionalStorageImpl.h]: better names would be "cheap replay" vs "exact replay"

File: aten/src/ATen/FunctionalTensorWrapper.cpp [hunk 7]
```diff
@@ -363,23 +371,25 @@ void FunctionalTensorWrapper::sync_() {
     return;
   }
   apply_updates();
-  regenerate_from_base();
+  regenerate_from_base(/*single_output_replay=*/true);
```
Comment (zou3519) [inline aten/src/ATen/FunctionalTensorWrapper.cpp]: so you're basically saying sync can only be used for internal replays and external replays need to use `_unwrap_functional_tensor`. I don't know if this is an existing invariant

Comment (zou3519) [reply aten/src/ATen/FunctionalTensorWrapper.cpp]: so I think this is a little brittle, but I'm not sure what to do about it...

File: torch/_functorch/eager_transforms.py (hunk 5 above)
Comment (anijain2305) [reply torch/_functorch/eager_transforms.py:1590]: Brought it back ... Codex and then no test failing convinced me that this was not needed. But, this is also harmless for compile. We can tackle this in a separate PR.

File: aten/src/ATen/FunctionalStorageImpl.h (hunk 6 above)
Comment (anijain2305) [reply aten/src/ATen/FunctionalStorageImpl.h]: changed

File: aten/src/ATen/FunctionalStorageImpl.h [hunk 8]
```diff
@@ -107,6 +107,21 @@ struct ViewMeta {
   virtual Tensor forward(const Tensor& base) = 0;
   virtual Tensor reverse(const Tensor& base, const Tensor& mutated_view) = 0;
 
+  // The cheap replay. Same tensor as `forward`, the exact replay, but a
```
Comment (zou3519) [inline aten/src/ATen/FunctionalStorageImpl.h]: I thought about this some more. Here's what I think a better design would be:
- don't try to call split()/unbind()/etc. always just replay as single output
- if is_multi_output_view_, during runtime, we set CreationMeta::MULTI_OUTPUT_NODE on the output tensors

though probably want to check with @soulitzer if this will run us into any problems. As far as I can tell CreationMeta::MULTI_OUTPUT_NODE  just controls autograd's erroring out on view+inplace

Comment (soulitzer) [reply aten/src/ATen/FunctionalStorageImpl.h]: Yep that's only used for erroring out on view+inplace. I don't think it would cause any problems, although the error message might now say something like "SelectBackward0 .... function that returns multiple views" ?probably not a huge deal though.

**Review by zou3519 — APPROVED** [review]

**Comment by zou3519** [comment]
minor bc-break for changing the IR, but maybe that doesn't count

**Comment by anmyachev** [comment]
We also observe regression in timm_models due to this change: https://github.com/intel/intel-xpu-backend-for-triton/issues/8077#issuecomment-5713356062
